core/editor/src/input/mouse.rs

```diff
@@ -1,5 +1,79 @@
 use bitflags::bitflags;
 
+// A position on the infinite canvas
+#[derive(Debug, Copy, Clone, Default, PartialEq)]
+pub struct DocumentPosition {
+	pub x: f64,
+	pub y: f64,
+}
+
+impl DocumentPosition {
+	pub fn distance(&self, other: &Self) -> f64 {
+		let x_diff = other.x - self.x;
+		let y_diff = other.y - self.y;
+		f64::sqrt(x_diff * x_diff + y_diff * y_diff)
+	}
+	pub fn rotate(&self, theta: f64) -> Self {
+		let cosine = theta.cos();
+		let sine = theta.sin();
+		Self {
+			x: self.x * cosine - self.y * sine,
+			y: self.x * sine + self.y * cosine,
+		}
+	}
+	pub fn add(&self, x: f64, y: f64) -> Self {
+		Self { x: self.x + x, y: self.y + y }
+	}
+	pub fn multiply(&self, x: f64) -> Self {
+		Self { x: self.x * x, y: self.y * x }
+	}
+}
+impl From<DocumentPosition> for (f64, f64) {
+	fn from(item: DocumentPosition) -> Self {
+		(item.x, item.y)
+	}
+}
+
+// The location of the viewport (or anything else) in the canvas
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct DocumentTransform {
+	pub location: DocumentPosition,
+	pub radians: f64,
+	pub degrees: f64,
+	pub scale: f64,
+	pub size: ViewportPosition,
+}
+
+impl Default for DocumentTransform {
+	fn default() -> Self {
+		Self {
+			location: DocumentPosition { x: 800., y: 800. },
+			scale: 2.4,
+			radians: 0.785398163, // 45 degrees in radians
+			degrees: 45.,
+			size: ViewportPosition::default(),
+		}
+	}
+}
+
+impl DocumentTransform {
+	pub fn transform_string(&self) -> String {
+		let inverse_scale = 1. / self.scale;
+		let size = DocumentPosition {
```

Comment:
If this represents the center of the viewport in viewport-coordinates it should be a ViewportPositon.